src/journal/src/Journal/Internal/ByteBuffer.hs

```diff
@@ -257,6 +266,36 @@ putBytes src dest = do
 getBytes :: ByteBuffer -> Int -> Int -> IO [Word8]
 getBytes bb offset len = undefined
 
+putByteString :: ByteBuffer -> BS.ByteString -> IO ()
+putByteString bb bs = do
+  let (fptr, I# offset#, I# len#) = BS.toForeignPtr bs
+  boundCheck bb (I# (len# -# 1#))
+  withForeignPtr fptr $ \(Ptr addr#) -> IO $ \s ->
+    case copyAddrToByteArray# addr# (bbData bb) offset# len# s of
+      s' -> (# s', () #)
+
+putLazyByteString :: ByteBuffer -> LBS.ByteString -> IO ()
+putLazyByteString bb lbs = do
+  let (fptr, I# offset#, I# len#) = BS.toForeignPtr (LBS.toStrict lbs)
+  boundCheck bb (I# (len# -# 1#))
+  withForeignPtr fptr $ \(Ptr addr#) -> IO $ \s ->
+    case copyAddrToByteArray# addr# (bbData bb) offset# len# s of
+      s' -> (# s', () #)
+
+getByteString :: ByteBuffer -> Int -> IO BS.ByteString
+getByteString bb len = do
+  bytes <- replicateM len (getByte bb)
+  return (BS.packBytes bytes)
```

Comment:
In the future we probably want to use `memcpy` for this. Like they do in `ByteString` for `copy` https://hackage.haskell.org/package/bytestring-0.11.2.0/docs/src/Data.ByteString.Internal.html#memcpy

`memcpy` should copy more bytes at the time, rather than going one by one.

Reply:
Fixed (using `copyMutableByteArrayToAddr#`), thanks!